bgpd/bgp_zebra.c

```diff
@@ -2316,6 +2316,102 @@ static int iptable_notify_owner(ZAPI_CALLBACK_ARGS)
 	return 0;
 }
 
+/* Process route notification messages from RIB */
+static int bgp_zebra_route_notify_owner(int command, struct zclient *zclient,
+					zebra_size_t length, vrf_id_t vrf_id)
+{
+	struct prefix p;
+	enum zapi_route_notify_owner note;
+	uint32_t table_id;
+	char buf[PREFIX_STRLEN];
+	afi_t afi;
+	safi_t safi;
+	struct bgp_node *rn;
+	struct bgp *bgp;
+	struct bgp_path_info *pi, *new_select;
+
+	if (!zapi_route_notify_decode(zclient->ibuf, &p, &table_id, &note,
+				      &afi, &safi)) {
+		zlog_err("%s : error in msg decode", __PRETTY_FUNCTION__);
+		return -1;
+	}
+
+	/* Get the bgp instance */
+	bgp = bgp_lookup_by_vrf_id(vrf_id);
+	if (!bgp) {
+		flog_err(EC_BGP_INVALID_BGP_INSTANCE,
+				"%s : bgp instance not found vrf %d",
+				__PRETTY_FUNCTION__, vrf_id);
+		return -1;
+	}
+
+	if ((afi < AFI_IP) || (afi >= AFI_MAX) ||
+	    (safi < SAFI_UNICAST) || (safi >= SAFI_MAX)) {
+		flog_err(EC_BGP_INVALID_ROUTE,
+			 "%s : invalid afi %d, safi %d", __PRETTY_FUNCTION__,
+			 afi, safi);
+	}
+
+	if (BGP_DEBUG(zebra, ZEBRA))
+		prefix2str(&p, buf, sizeof(buf));
+
+	/* Find the bgp route node */
+	rn = bgp_afi_node_lookup(bgp->rib[afi][safi], afi, safi, &p,
+				 &bgp->vrf_prd);
+	if (!rn) {
+		flog_err(EC_BGP_INVALID_ROUTE,
+			 "%s : route %s not found afi %d, safi %d",
+			 __PRETTY_FUNCTION__, buf, afi, safi);
+		return -1;
+	}
+	bgp_unlock_node(rn);
+
+	switch (note) {
+	case ZAPI_ROUTE_FAIL_INSTALL:
+		UNSET_FLAG(rn->flags, BGP_NODE_FIB_INSTALLED);
+		flog_err(EC_BGP_FIB_INSTALL_ERROR,
```

Comment:
If the failure is happening in the rib, and the error is reported there, do we really need *another* error about this situation?

Reply:
Additionally do we need to unset the install pending flag here?

Reply:
Pending flag will remain set till route is successfully installed

Reply:
> If the failure is happening in the rib, and the error is reported there, do we really need _another_ error about this situation?

@kssoman can you please address this comment?

Reply:
Will remove this log as logs are added in zebra_rib.c

Reply:
changes done